builtin/providers/terraform/provider.go

```diff
@@ -1,21 +1,151 @@
+	"fmt"
+
+	"github.com/hashicorp/terraform/config/configschema"
+	"github.com/hashicorp/terraform/providers"
+	"github.com/zclconf/go-cty/cty"
 )
 
-// Provider returns a terraform.ResourceProvider.
-func Provider() terraform.ResourceProvider {
-	return &schema.Provider{
-		ResourcesMap: map[string]*schema.Resource{
-			"terraform_remote_state": schema.DataSourceResourceShim(
-				"terraform_remote_state",
-				dataSourceRemoteState(),
-			),
+// Provider is an implementation of providers.Interface
+type Provider struct {
+	// Provider is the schema for the provider itself.
+	Schema providers.Schema
+
+	// DataSources maps the data source name to that data source's schema.
+	DataSources map[string]providers.Schema
+}
+
+// NewProvider returns a new terraform provider
+func NewProvider() *Provider {
+	return &Provider{}
+}
+
+// GetSchema returns the complete schema for the provider.
+func (p *Provider) GetSchema() providers.GetSchemaResponse {
+	return providers.GetSchemaResponse{
+		Provider: providers.Schema{
+			Version: 1,
 		},
-		DataSourcesMap: map[string]*schema.Resource{
-			"terraform_remote_state": dataSourceRemoteState(),
+		DataSources: map[string]providers.Schema{
+			"terraform_remote_state": {
+				Block: &configschema.Block{
+					Attributes: map[string]*configschema.Attribute{
+						"backend": {
+							Type:     cty.String,
+							Required: true,
+						},
+						"config": {
+							Type:     cty.DynamicPseudoType,
+							Optional: true,
+						},
+						"defaults": {
+							Type:     cty.DynamicPseudoType,
+							Optional: true,
+						},
+						"outputs": {
+							Type:     cty.DynamicPseudoType,
+							Computed: true,
+						},
+						"workspace": {
+							Type:     cty.String,
+							Optional: true,
+						},
+					},
+				},
+			},
 		},
 	}
 }
+
+// ValidateProviderConfig is used to validate the configuration values.
+func (p *Provider) ValidateProviderConfig(providers.ValidateProviderConfigRequest) providers.ValidateProviderConfigResponse {
+	// At this moment there is nothing to configure for the terraform provider,
+	// so we will happily return without taking any action
+	var res providers.ValidateProviderConfigResponse
+	return res
+}
+
+// ValidateDataSourceConfig is used to validate the data source configuration values.
+func (p *Provider) ValidateDataSourceConfig(providers.ValidateDataSourceConfigRequest) providers.ValidateDataSourceConfigResponse {
+	// At this moment there is nothing to configure for the terraform provider,
+	// so we will happily return without taking any action
+	var res providers.ValidateDataSourceConfigResponse
+	return res
+}
+
+// Configure configures and initializes the provider.
+func (p *Provider) Configure(providers.ConfigureRequest) providers.ConfigureResponse {
+	// At this moment there is nothing to configure for the terraform provider,
+	// so we will happily return without taking any action
+	var res providers.ConfigureResponse
+	return res
```

Comment:
We don't need to do this for the first pass here, since what you've written here matches what the old codepath did, but in a subsequent commit it'd be nice to move the backend configuration validate call that's currently inside the read method into here so that we can catch provider configuration errors in `terraform validate` as well as during `terraform plan`.

With that said, before we go down that road we should have a think about how to deal with potential unknown values in the config here, since backends generally do not expect to deal with those. Probably what we'd do here is test to see if the configuration value is "wholly known" and skip validating it here if it isn't, and then retain the validation during the read function too so we're still guaranteed to validate at least once but we can opportunistically validate sooner in the simple case.

Reply:
I'm concerned about forgetting this feedback - would you like me to implement what you've suggested _now_ so we have a placeholder that we can potentially adjust later? Or are you confident that it'll "come up" again?

Reply:
Looks like I actually left this feedback in the wrong place anyway... should've been in the `ValidateDataSource` method. :roll_eyes:

Since this is just polish and we could easily ship without it, I wouldn't prioritize this over other work. It could be fine to just leave a FIXME note in the `ValidateDataSource` method about it and we can catch it on a future pass.

I'd rather merge this without it since what you've done here represents an as-close-as-possible port of what was there before, which is a great baseline to have here before we start trying to improve on the prior functionality.